### PR TITLE
Log cosh

### PR DESCRIPTION
**Pull request by ven-k**
- Log-Cosh gives losses similar to MSE, but isn't much affected by occasional wild predictions. 
- It can be differentiated twice
- Keras has similar implementation.
- It's computed by 
`log_cosh = sum (log (cosh (yhat - y))) / n`

### CODE REVIEW

File: src/measures/continuous.jl [hunk 1]
```diff
@@ -320,3 +320,33 @@ function (m::MAPE)(ŷ::Vec{<:Real}, y::Vec{<:Real})
...
+"""
+const log_cosh = LogCosh()
+
+metadata_measure(LogCosh;
+    name                     = "log_cosh",
+    target_scitype           = Union{Vec{Continuous},Vec{Count}},
+    prediction_type          = :deterministic,
+    orientation              = :loss,
+    reports_each_observation = false,
+    is_feature_dependent     = false,
```
Comment (ablaom) [inline src/measures/continuous.jl:345]: Suggestion (see **C** below): `reports_each_observation = true`.

File: src/measures/continuous.jl [hunk 2]
```diff
@@ -320,3 +320,33 @@ function (m::MAPE)(ŷ::Vec{<:Real}, y::Vec{<:Real})
...
+    supports_weights         = false,
+    docstring                = "LogCosh; aliases: `log_cosh`.")
+
+function (log_cosh::LogCosh)(ŷ::Vec{<:Real}, y::Vec{<:Real})
+    check_dimensions(ŷ, y)
+    return sum(log.(cosh.(ŷ-y))) / length(y)
+end
```
Comment (ablaom) [inline src/measures/continuous.jl:352]: Suggestion (see **C** below):  return a vector of per-observation losses: `return log.(cosh.(ŷ-y))) / length(y)`

File: test/measures/continuous.jl [hunk 3]
```diff
@@ -12,6 +12,7 @@ rng = StableRNG(666899)
     @test isapprox(mean(l1(yhat, y, w)), mae(yhat, y, w))
     @test isapprox(mean(l2(yhat, y)), 5)
     @test isapprox(mean(l2(yhat, y, w)), rms(yhat, y, w)^2)
+    @test isapprox(log_cosh(yhat, y), 1.3715546675)
```
Comment (ablaom) [inline test/measures/continuous.jl:16]: Modify test accordingly.

**Review by ablaom — COMMENTED** [review]
@ven-k Thanks for this contribution and for taking the time to wrap your head around the API!

**C**

As the loss is just the mean of a per-observation loss, can I suggest the changes above? By default this loss will be aggregated in resampling using the mean (`aggregation(...) = Mean()`). In this way the per-observation losses will be available, for example, to hyper-parameter optimisation strategies that use this information (some Bayesian methods). 

So then the code will look basically the same as `l1` which has this behaviour:

```julia
julia> l1([1,2,3], [2,3,4])
3-element Array{Int64,1}:
 1
 1
 1
```

File: src/measures/continuous.jl (hunk 1 above)
Comment (OkonSamuel) [reply src/measures/continuous.jl:345]: are we sure he needs `reports_each_observation = true` since he said the `Log-Cosh` loss is similar to `rms` and because of the definition in his comment as `log_cosh = sum (log (cosh (yhat - y))) / n`. 
To me it looks like he meant to set `reports_each_observation = false`

File: src/measures/continuous.jl (hunk 2 above)
Comment (OkonSamuel) [reply src/measures/continuous.jl:352]: The `sum` is missing in this definition

File: src/measures/continuous.jl (hunk 1 above)
Comment (ven-k) [reply src/measures/continuous.jl:345]: @OkonSamuel  `mean(log(cosh(yhat-y)))` returns value similar to MSE (it's what Tensorflow returns). But as @ablaom mentioned in the suggestion **C**, returning vector will help in tuning of hyperparameters, so I've modified accordingly. 
PS: The definition is `Log_Cosh(x) = log(cosh(x))`

Comment (ven-k) [reply src/measures/continuous.jl:345]: @OkonSamuel  `mean(log(cosh(yhat-y)))` returns value similar to MSE (it's what Tensorflow returns). But as @ablaom mentioned in the suggestion **C**, returning vector will help in tuning of hyperparameters, so I've modified accordingly. 
PS: The definition is `Log_Cosh(x) = log(cosh(x))`

Comment (OkonSamuel) [reply src/measures/continuous.jl:345]: Oh. I get you now

**Review by OkonSamuel — APPROVED** [review]

File: src/measures/continuous.jl [hunk 4]
```diff
@@ -320,3 +320,33 @@ function (m::MAPE)(ŷ::Vec{<:Real}, y::Vec{<:Real})
...
 end
+
+struct LogCosh <: Measure end
+
+"""
+    LogCosh(ŷ, y)
+
+Log-Cosh loss:
+
+``\\text{Log-Cosh} = m^{-1}∑ᵢ log(cosh(ŷᵢ-yᵢ))``
```
[the remark this answers is not in the document]
Comment (ven-k) [reply src/measures/continuous.jl]: Done

**Review by ablaom — APPROVED** [review]
@OkonSamuel Yeah it is a bit confusing. Some measures that could be implemented as per-observation losses are not, eg, `mae` (which I guess essentially duplicates `l1`, which *is* per-observation loss). 

@ven-k  For the record, measures is overdue for a thorough review and  migration out of MLJ https://github.com/alan-turing-institute/MLJBase.jl/issues/299 .

**Comment by ablaom** [comment]
@ven-k  Now live (see above)

**Comment by ven-k** [comment]
Thanks! @ablaom and @OkonSamuel 

I am following #299 and #416. I am interested in StatisticalMeasures and possible visualisation package.
